File: python/complex-numbers/complex_numbers.py
"""Implement complex numbers."""

import math
import numbers

class ComplexNumber:
    """Complex Number."""

    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary

    def __eq__(self, other):
        if isinstance(other, numbers.Real):
            return self == ComplexNumber(other, 0)

        return (
            isinstance(other, ComplexNumber)
            and self.real == other.real
            and self.imaginary == other.imaginary)

    def __add__(self, other):
        if isinstance(other, numbers.Real):
            return self + ComplexNumber(other, 0)

        return ComplexNumber(self.real + other.real, self.imaginary + other.imaginary)

    def __mul__(self, other):
        if isinstance(other, numbers.Real):
            return self * ComplexNumber(other, 0)

        return ComplexNumber(
            self.real * other.real - self.imaginary * other.imaginary,
            self.imaginary * other.real + self.real * other.imaginary
        )

    def __sub__(self, other):
        if isinstance(other, numbers.Real):
            return self - ComplexNumber(other, 0)

        return ComplexNumber(self.real - other.real, self.imaginary - other.imaginary)

    def __radd__(self, other):
        return self + other

    def __rmul__(self, other):
        return self * other

    def __rsub__(self, other):
        return other + -1 * self

    def __rtruediv__(self, other):
        div = (self.real**2 + self.imaginary**2)
        return other * ComplexNumber(self.real / div, - self.imaginary / div)

    def __truediv__(self, other):
        if isinstance(other, numbers.Real):
            return self / ComplexNumber(other, 0)

        div = (other.real**2 + other.imaginary**2)
        return ComplexNumber(
            (self.real * other.real + self.imaginary * other.imaginary) / div,
            (self.imaginary * other.real - self.real * other.imaginary) / div,
        )

    def __abs__(self):
        return math.sqrt(self.real ** 2 + self.imaginary ** 2)

    def conjugate(self):
        """Return the conjugate."""
        return ComplexNumber(self.real, -self.imaginary)

    def exp(self):
        """Return the exponent."""
        e_i = ComplexNumber(math.cos(self.imaginary), math.sin(self.imaginary))
        return e_i * math.exp(self.real)

    def __repr__(self):
        return f"{self.real} + {self.imaginary} * i"


# vim:ts=2:sw=2:expandtab

File: go/complex-numbers/complex_numbers.go
// Package complex provides complex number support.
package complex

import "math"

// Number is a complex number.
type Number struct {
	real float64
	imag float64
}

// Real gives the real part of a complex number.
func (n Number) Real() float64 {
	return n.real
}

// Imaginary gives the imaginary part of a complex number.
func (n Number) Imaginary() float64 {
	return n.imag
}

// Add returns the sum of two complex numbers.
func (n Number) Add(other Number) Number {
	return Number{n.real + other.real, n.imag + other.imag}
}

// Subtract returns the difference of two complex numbers.
func (n Number) Subtract(other Number) Number {
	return Number{n.real - other.real, n.imag - other.imag}
}

// Multiply returns the product of two complex numbers.
func (n Number) Multiply(other Number) Number {
	return Number{
		n.real*other.real - n.imag*other.imag,
		n.imag*other.real + n.real*other.imag,
	}
}

// Times returns the product of a complex and real number.
func (n Number) Times(factor float64) Number {
	return Number{factor * n.real, factor * n.imag}
}

// Div returns the quotient of a complex and real number.
func (n Number) Div(factor float64) Number {
	return n.Times(1 / factor)
}

func (n Number) absSqr() float64 {
	return n.real*n.real + n.imag*n.imag
}

// Divide returns the quotient of two complex numbers.
func (n Number) Divide(other Number) Number {
	return n.Multiply(other.Conjugate()).Div(other.absSqr())
}

// Conjugate returns the conjugate of a complex number.
func (n Number) Conjugate() Number {
	return Number{n.real, -n.imag}
}

// Abs returns the absolute value of a complex number.
func (n Number) Abs() float64 {
	return math.Sqrt(n.absSqr())
}

// Exp returns e ** n for a complex number n.
func (n Number) Exp() Number {
	return Number{math.Cos(n.imag), math.Sin(n.imag)}.Times(math.Exp(n.real))
}
